File: backend-api/scrapers/internshala.js
const puppeteer = require('puppeteer');

const scrapeInternshala = async (skills) => {
  const browser = await puppeteer.launch({
    headless: 'new',
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
  });

  const page = await browser.newPage();
  await page.setUserAgent(
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36'
  );

  // Set viewport to look like real browser
  await page.setViewport({ width: 1280, height: 800 });

  const skill = skills[0] || 'software';
  const url = `https://internshala.com/jobs/keywords-${skill.toLowerCase().replace(/\s+/g, '-')}-jobs/`;
  console.log('Scraping:', url);

  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    // Wait for job cards to appear
    await page.waitForSelector('.internship_meta, .individual_internship', { timeout: 10000 });
  } catch (e) {
    console.log('Page load issue:', e.message);
  }

  // Log page HTML for debugging
  const html = await page.content();
  console.log('Page HTML length:', html.length);

  const jobs = await page.evaluate(() => {
    // Try multiple possible selectors
    const cards =
      document.querySelectorAll('.individual_internship') ||
      document.querySelectorAll('.internship-list-container .internship_meta');

    return Array.from(cards).slice(0, 15).map(card => {
      const titleEl = card.querySelector('.job-internship-name, .profile, h3 a');
      const companyEl = card.querySelector('.company-name, .company_name');
      const locationEl = card.querySelector('.location_link, .locations span, .location');
      const salaryEl = card.querySelector('.stipend, .salary');
      const linkEl = card.querySelector('a.job-title-href, a[href*="/job-detail/"]');

      return {
        title: titleEl?.innerText?.trim(),
        company: companyEl?.innerText?.trim(),
        location: locationEl?.innerText?.trim() || 'India',
        salary: salaryEl?.innerText?.trim(),
        applyUrl: linkEl
          ? 'https://internshala.com' + linkEl.getAttribute('href')
          : null,
        source: 'Internshala'
      };
    }).filter(j => j.title && j.applyUrl);
  });

  console.log('Scraped jobs count:', jobs.length);
  await browser.close();
  return jobs;
};

module.exports = { scrapeInternshala };

File: backend-api/app/scrapers/internshala.py
from __future__ import annotations

from urllib.parse import quote

import httpx
from bs4 import BeautifulSoup

from app.schemas.scraped_job import ScrapedJob
from app.scrapers.base import BaseScraper


class InternshalaScraper(BaseScraper):
    source = "Internshala"

    async def scrape(self, query: str) -> list[ScrapedJob]:
        normalized = (query or "software developer").strip().lower().replace(" ", "-")
        url = f"https://internshala.com/jobs/keywords-{quote(normalized)}-jobs/"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        }

        async with httpx.AsyncClient(timeout=30.0, follow_redirects=True) as client:
            res = await client.get(url, headers=headers)
            res.raise_for_status()

        soup = BeautifulSoup(res.text, "html.parser")
        cards = soup.select(".individual_internship")[:15]
        jobs: list[ScrapedJob] = []
        for card in cards:
            title_el = card.select_one(".job-internship-name") or card.select_one(".profile")
            company_el = card.select_one(".company-name")
            location_el = card.select_one(".location_link")
            salary_el = card.select_one(".stipend")
            link_el = card.select_one("a.job-title-href")

            href = link_el.get("href") if link_el else None
            if not title_el or not href:
                continue
            apply_url = href if href.startswith("http") else f"https://internshala.com{href}"

            jobs.append(
                ScrapedJob(
                    title=title_el.get_text(strip=True),
                    company=(company_el.get_text(strip=True) if company_el else "Unknown"),
                    location=(location_el.get_text(strip=True) if location_el else "India"),
                    salary=(salary_el.get_text(strip=True) if salary_el else None),
                    applyUrl=apply_url,
                    source=self.source,
                    employmentType="full-time",
                )
            )
        return jobs
